{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "title: Scikick for Single Cell Analysis\n",
    "---\n",
    "\n",
    "The following repository is an interactively executed demonstration of Scikick usage through an analysis of single cell transcriptomic (scRNAseq) data. \n",
    "\n",
    "This analysis is based on work from the [Orchestrating Single Cell Analysis](http://bioconductor.org/books/release/OSCA/) book.\n",
    "\n",
    "The `notebooks` directory contains a series of notebooks which were developed to analyze scRNAseq datasets. Each notebook will be introduced and added to the Scikick project as if they were developed throughout a true project timeline.  \n",
    "\n",
    "The final report generated by Scikick for this tutorial project can be seen [here](../../single-cell_analysis/report/out_html/index.html).\n",
    "\n",
    "Links to `sk <command> --help` outputs are provided throughout this tutorial for review and convenience. It is recommended to first read at least the [\"Hello World\"](hello_world.html) usage of Scikick prior to reading this page. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preparing the Software Environment\n",
    "\n",
    "Essential screenshots of the report state are provided throughout this tutorial and there is no need to directly execute the commands. However, to interactively view the state of the report at each stage of the tutorial, commands must be executed interactively in-line with the text. Full execution of this notebook takes approximately 10 minutes and installation of the required software for the project can take up to a couple of hours.\n",
    "\n",
    "Later in this tutorial, a demonstration will show how to execute this project using [Singularity](https://sylabs.io/docs/) with Scikick such that no project-specific dependencies require installation. \n",
    "\n",
    "If you are executing the tutorial commands interactively, you should ensure you have followed the setup instructions in this section. \n",
    "\n",
    "If you are reading only, continue to the [initialize scikick](#Initialize-Scikick) section."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Executing Tutorial Commands\n",
    "\n",
    "This tutorial utilizes a [bash kernel](https://pypi.org/project/bash_kernel/) to execute bash commands in Jupyter. Additionally, this tutorial was developed in [Jupyter Lab](https://jupyter.org/install); when viewing the report site throughout the tutorial, it may be necessary to click \"trust html\" to enable all site content to be properly displayed.\n",
    "\n",
    "Commands can also simply be executed in a shell terminal and reports viewed by opening in any modern web browser."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Install Scikick\n",
    "\n",
    "Ensure you have a working installation of [Scikick](https://github.com/matthewcarlucci/scikick)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Obtain Project Notebooks\n",
    "\n",
    "To execute the code in this tutorial, download [Scikick's source code repository](https://github.com/matthewcarlucci/scikick) and navigate to the directory containing the analysis code and this Jupyter notebook found under `docs/scikick_documentation/`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Install Project Dependencies\n",
    "\n",
    "The software used throughout the analysis is required for execution. This project uses [R >= 4.0](https://www.r-project.org/) for analysis. Dependencies for the analysis can be installed from the project root (*i.e.* `single-cell_analysis/`) in R with:\n",
    "\n",
    "```r\n",
    "# Install BiocManager and remotes prior to this command\n",
    "BiocManager::install(remotes::local_package_deps(dependencies=TRUE), version=\"3.12\")\n",
    "```\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Initialize Scikick"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Starting from scikick source code docs/scikick_documentation (this notebook)\n",
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Move into the scRNAseq project\n",
    "cd single-cell_analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div hidden>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# (this code cell is hidden by HTML tags) - Remove previous executions of this tutorial\n",
    "if [ -f scikick.yml ]; then\n",
    "    rm scikick.yml || echo \"Not found\"\n",
    "    rm -rf report || echo \"Not found\"\n",
    "    rm -rf output || echo \"Not found\"\n",
    "    # If tutorial is executed some notebooks would have been moved\n",
    "    mv notebooks/nestorowa/* notebooks/ || echo \"No notebooks found\"\n",
    "fi"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the project root directory, the Scikick project is initiated with [sk init](help.html#init)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sk init -y"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will add some optional website styling to the configuration file (`scikick.yml`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cat _site.yml \n",
    "cat _site.yml >> scikick.yml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Adding to the Homepage\n",
    "\n",
    "It is often useful to first start with writing some background on the project and to provide an overview of data sources, analysis goals, and the current state of the project. The `index.Rmd` notebook provides an overview of the scRNAseq analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Inspecting index.Rmd contents\n",
    "cat index.Rmd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`index.Rmd` is added with [sk add](help.html#add)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sk add index.Rmd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`index.Rmd` is the only special file name in Scikick projects. It will be used for homepage content. We can see in the message above that Scikick recognized this.\n",
    "\n",
    "An initial website will now be created with [sk run](help.html#run)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sk run"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "caption": "somecaption",
    "label": "fig:somelabel"
   },
   "source": [
    "---\n",
    "\n",
    "*Note: The console outputs seen above for project map, site layout, and html generation will be supressed throughout the rest of the tutorial by using `-q` with `sk run`.*\n",
    "\n",
    "We can see the contents of `index.Rmd` were added to the homepage. \n",
    "\n",
    "[![](../../single-cell_analysis/imgs/homepage.png)](../../single-cell_analysis/report/out_html/index.html)\n",
    "\n",
    "Now we are ready to start adding analysis to the project.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Nestorowa Dataset Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Modular Notebooks\n",
    "\n",
    "During data exploration it is often a good idea to develop analyses in separate notebooks when multiple stages of data transformation are being performed such that notebooks:\n",
    "\n",
    "1. Remain focused on a single topic.\n",
    "2. Each have manageable namespaces and resource usage for data objects. \n",
    "3. Maintain clear checkpoints in the state of main data objects.\n",
    "\n",
    "To practice this modularized approach, notebooks in this tutorial are small and focus on specific tasks. Scikick provides tools to manage notebooks developed in this modular fashion. Modularization may be slightly exagerated in this project where a typical project may combine many of these steps into a single notebook at the user's discretion."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Importing Data\n",
    "\n",
    "The `notebooks/import.Rmd` notebook was developed to import the scRNAseq dataset from *Nestorowa et al 2016*. This notebook:\n",
    "\n",
    "- Describes further details on where the data comes from.\n",
    "- Makes a few adjustments to the data to make it easier to work with.\n",
    "- Inspects the object after it has finished downloading.\n",
    "\n",
    "The data is then saved to a file for later usage."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cat notebooks/import.Rmd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The notebook will now be added to the project."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sk add notebooks/import.Rmd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Checking the new state of the project with [sk status](help.html#status)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sk status"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Scikick has determined that `import.Rmd` must be executed since it is missing its output report file and the `index.Rmd` only requires small changes to its final page output (to include a menu item for `import.Rmd`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sk run"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "We can see that `import.Rmd` was added to the report site (under the navigation bar as \"Import\") with no additional configuration necessary.\n",
    "\n",
    "[![](../../single-cell_analysis/imgs/import.png)](../../single-cell_analysis/report/out_html/notebooks/nestorowa/import.html)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Quality Control\n",
    "\n",
    "A notebook was developed which performs quality control on the data downloaded by the `import.Rmd` notebook. That is, `quality_control.Rmd` must be executed after `import.Rmd`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cat notebooks/quality_control.Rmd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`sk add` with the `-d/--depends-on` flag is used to configure the execution order of these two notebooks. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sk add notebooks/quality_control.Rmd --depends-on notebooks/import.Rmd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sk status"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "However, since `import.Rmd` has already run, only `quality_control.Rmd` requires execution."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sk run -q"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[![](../../single-cell_analysis/imgs/quality_control.png)](../../single-cell_analysis/report/out_html/notebooks/nestorowa/quality_control.html)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Inspecting the Project Map\n",
    "\n",
    "With multiple notebooks present in the final report, the order of execution is now no longer immediately clear. Viewing the project maps generated by Scikick at the bottom of each page rectifys this by clearly outlining the connection made between `quality_control` and `import` (with `-d/--depends-on`). This map can also be used to navigate across pages of the report.\n",
    "\n",
    "[![](../../single-cell_analysis/imgs/quality_control_project_map.png)](../../single-cell_analysis/report/out_html/notebooks/nestorowa/quality_control.html)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Normalization \n",
    "\n",
    "A notebook is added for implementing normalization and variance modeling of the transcript counts for the samples which survived quality control in `quality_control.Rmd`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cat notebooks/normalization.Rmd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sk add notebooks/normalization.Rmd -d notebooks/quality_control.Rmd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sk run -q"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have now obtained a meaningful dataset for exploring biological results."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Further Exploration of the Nestorowa Data\n",
    "\n",
    "Once the data has been cleaned and normalized into a meaningful format for interprettation, it is common to perform some exploration of the data. Some common scRNAseq data exploration tasks are performed in a `further_exploration.Rmd` notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sk add notebooks/further_exploration.Rmd -d notebooks/normalization.Rmd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sk run -q"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Nestorowa Analysis Summary\n",
    "\n",
    "We now have a complete analysis of the Nestorowa dataset. Each notebook's respective page is found in the Scikick report. The project map clearly shows the order of execution of these pages. \n",
    "\n",
    "[![](../../single-cell_analysis/imgs/nestorowa_final.png)](../../single-cell_analysis/report/out_html/notebooks/nestorowa/further_exploration.html)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With Scikick usage throughout this project thus far, we can now be confident that, when we remove the report directory, we are able to regenerate all results with `sk run`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# E.g. We could remove the report and regenerate from scratch.\n",
    "# This will be skipped to save computations.\n",
    "# rm -rf report\n",
    "# sk run -q"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can take a look at the files required to re-execute this project."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ls notebooks/*Rmd\n",
    "ls scikick.yml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now come back and check on the state of this project at any time in the future to find that there are no pending updates and that the report in `report/out_html` represents a full execution of these notebooks."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# View notebooks, dependencies, and state\n",
    "sk status -v"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Updating the Dataset\n",
    "\n",
    "If we now update the original data import (e.g. modify `import.Rmd` to use a new version of the raw data), Scikick is aware of the state and relationship between multiple notebooks. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using touch to simulate an update to import.Rmd\n",
    "touch notebooks/import.Rmd\n",
    "sk status"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We see that `import` (`s--` indicating an update to \"self\") and all downstream notebooks (`-e-` indicating an update to an external dependency) must be re-executed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sk run -q"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Additional Experiments\n",
    "\n",
    "Scientific projects often develop as a series of experiments where analsyis is performed in stages. We will now add two similar parallel workflows for two new scRNAseq experiments.\n",
    "\n",
    "## Reorganizing Content\n",
    "\n",
    "As a project moves forward, more datasets may be added requiring reorganization of the project. Workflow configurations can be difficult to coordinate with file re-organization. Scikick provides features to accomodate this. The [sk mv](help.html#move) command enables dynamic reorganization of projects while attempting to minimize the need for re-execution. It does this by applying a standard shell `mv` command (or `git mv` if using `-g`), while also moving the cooresponding report files such that re-execution is not necessary.\n",
    "\n",
    "For this project, we will use `sk mv` to migrate all analysis of the Nestorowa dataset above to a subfolder for this experiment such that new experiemnts can be added in a parallel fashion. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir notebooks/nestorowa\n",
    "sk mv notebooks/*.Rmd notebooks/nestorowa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that no code requires re-execution."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sk status"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add New Notebook Collections\n",
    "\n",
    "A similar series of notebooks were developed for the two additional experiments."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sk add notebooks/grun/import.Rmd\n",
    "sk add notebooks/grun/quality_control.Rmd -d notebooks/grun/import.Rmd\n",
    "sk add notebooks/grun/normalization.Rmd -d notebooks/grun/quality_control.Rmd\n",
    "sk add notebooks/grun/further_exploration.Rmd -d notebooks/grun/normalization.Rmd\n",
    "\n",
    "sk add notebooks/paul/import.Rmd\n",
    "sk add notebooks/paul/quality_control.Rmd -d notebooks/paul/import.Rmd\n",
    "sk add notebooks/paul/normalization.Rmd -d notebooks/paul/quality_control.Rmd\n",
    "sk add notebooks/paul/further_exploration.Rmd -d notebooks/paul/normalization.Rmd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merging Experiments\n",
    "\n",
    "Finally, a set of notebooks are added which perform a combined analysis of the three datasets that were each prepared in parallel. This `merge.Rmd` stage depends on the results of each of the data preparations performed previously (*i.e.,* `quality_control.Rmd` and `normalization.Rmd` notebooks), however, it does not depend on the biological analysis that was performed for each dataset (*i.e.,* the analyses performed in `further_exploration.Rmd` notebooks are not used by `merge.Rmd`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sk add notebooks/merged/merge.Rmd -d notebooks/grun/quality_control.Rmd -d notebooks/paul/quality_control.Rmd -d notebooks/nestorowa/normalization.Rmd\n",
    "sk add notebooks/merged/combined_analysis.Rmd -d notebooks/merged/merge.Rmd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Utilizing Parallelization\n",
    "\n",
    "With a parallel series of notebooks like in the current state of this project, execution of notebooks in parallel can be utilized with a flag (`-j8`) passed to snakemake with the `-s` [sk run](help.html#run) flag.\n",
    "\n",
    "The two additional experiments and merged analysis will now be executed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sk run -q -s -j8"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And a well organized final report site is generated. The project map in the report now features branching sets of analyses.\n",
    "\n",
    "[![](../../single-cell_analysis/imgs/merged_project_map.png)](../../single-cell_analysis/report/out_html/notebooks/merged/combined_analaysis.html)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Summary\n",
    "\n",
    "This demonstration utilizes real datasets and real analyses to demonstrate how Scikick is used in a practical setting when adapting to new analysis additions. Use of Scikick in a less controlled setting (when the analysis is not predetermined) should present even further utility to maintain workflow connections and reporting capabilities."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Appendix: Robust Archival"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Utilizing the Snakemake Backend\n",
    "\n",
    "Scikick is implemented through snakemake workflows allowing for usage of many features of snakemake. Users familliar with snakemake usage may be able to take advantage of many flags while using Scikick. Some frequently used examples are provided here."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Forced re-execution\n",
    "\n",
    "Passing snakemake arguments with `-s` and utilizing the snakemake `-F` flag will force the entire project to execute from scratch. This can be a useful sanity check when time permits."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This will be skipped to save computations.\n",
    "# sk run -s -F"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Execution with Singularity\n",
    "\n",
    "When readers attempt to reproduce this work or any other computational work, they may run into issues with software or data. In these cases, it is useful for the reader to see: \n",
    "\n",
    "1. That a fully reproducible archive exists (i.e. that the work is reproducible if the exact compute environment is used).\n",
    "2. That, if needed, they could utilize this static software environment to reproduce the results.\n",
    "\n",
    "The OSCA project maintains a docker image at [Docker Hub](https://hub.docker.com/r/bioconductor/orchestratingsinglecellanalysis). This image can first be assigned to the Scikick project with"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using a fixed image tag for future reproducibility \n",
    "sk config --singularity docker://bioconductor/orchestratingsinglecellanalysis:RELEASE_3_12"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Providing the flag `--use-singularity` to snakemake, will download the container and execute all notebooks within this container."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Automated Re-execution (i.e. Continuous Analysis)\n",
    "\n",
    "Configurations of singularity like the above effectively makes it possible to execute any Scikick data analysis project with only the core Scikick dependencies and no analysis software dependencies directly required.\n",
    "\n",
    "This scRNAseq project additionally requires the image to have write access to the `/home/cache` directory for data downloads and so a writeable location must be provided for execution with the additional argument `--singularity-args`.  Additionally, the script is configured to distribute notebook executions across a SLURM cluster. A short `run.sh` script contains a call to Scikick with these necessary arguments for full execution."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cat run.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Scikick Continuous Analysis with GitLab\n",
    "\n",
    "Linking this execution with a continuous integration service implements a version of [Continuous Analysis](https://greenelab.github.io/continuous_analysis/) where the archived reports may be refered to at any time in the future knowing that they are reproducible. A template for such a configuration using GitLab CI is provided below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cat .gitlab-ci.yml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The analysis codebase will now contain Scikick reports that are fully verified for reproducibility."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  },
  "toc-autonumbering": false,
  "toc-showcode": false,
  "toc-showmarkdowntxt": false,
  "toc-showtags": false
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
